### PR TITLE
add dist hooks support for custom device

### LINKED ISSUE #104389: distributed hooks want to support custom device
### 🐛 Describe the bug

for distributed hooks, there are some hard-codes, for cutsom device (privateuse1 backend, cuda-like device), we also want to support these hooks.

### Versions

main

### PR DESCRIPTION
**Pull request by heidongxianhua**
Fixes https://github.com/pytorch/pytorch/issues/104389
1. Now for distributed hooks, there are some hard-code like `cuda`, we want to support these hooks for custom device( privateuse1 backend), so we use the abstract device-module to run some funcs.
2. In `torch/nn/parallel/distributed.py`, I want to define a variable `self.device_module = getattr(torch, self.device_type, None)` so that we can reuse it, but it will cause an error in serialization, `TypeError: cannot pickle 'module' object`. So we call the func by `getattr` when needs.

### CODE REVIEW

**Comment by DerickZo** [comment]
您好，我已收到邮件，查阅后即刻回复^_^

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py [hunk 1]
```diff
@@ -38,7 +39,8 @@ def _reducer_allreduce_and_upcast_hook(
     fut = reducer._run_allreduce_hook(bucket)
     ret_fut = torch.futures.Future()
     stream = hook_state.upcast_stream
-    with torch.cuda.stream(stream):
+    device_module = _get_device_module(stream.device.type)
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:42]: This is a hot path that runs during gradient allreduce, so we should cache device_module look up when the hoook state is initialized.

File: torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py [hunk 2]
```diff
@@ -37,20 +38,22 @@ def _check_valid_functional_optim(self):
...
-    gradient_is_bucket_view: bool
+    gradient_is_bucket_view: bool,
+    device_type: str
 ) -> Callable[[Any, dist.GradBucket], torch.futures.Future[torch.Tensor]]:
     r"""
     If torch.distributed.optim._apply_optimizer_in_backward is used to overlap
     optimizer with backward pass, DDP will run the below hook to run optimizer
     step for parameters after gradient communication has taken place.
     """
+    device_module = _get_device_module(device_type)
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:54]: same comment as above.

File: torch/distributed/algorithms/model_averaging/utils.py [hunk 3]
```diff
@@ -31,8 +32,10 @@ def average_parameters(
     flat_params = torch.cat([p.data.reshape(-1) for p in params_it1])
     flat_params /= dist.get_world_size(group_to_use)
     # Make sure the allreduce will not conflict with any other ongoing process group.
-    if torch.cuda.is_available():
-        torch.cuda.synchronize()
+    device_type = dist.distributed_c10d._get_pg_default_device(group_to_use).type
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/utils.py:35]: this should probably be set at initialization time too.

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -703,6 +703,11 @@ def __init__(
             )
 
         self.device_type = list(distinct_device_types)[0]
+        # use `torch.cuda` module to solve `xla` device
+        self._device_module_type = (
+            self.device_type if self.device_type != "xla" else "cuda"
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:708]: Can you explain this? why do we want to set _device_module_type to "cuda" if self.device_type == xla?

Also in general, we should avoid harcode of specific device strings in application logic.

**Review by rohan-varma — COMMENTED** [review]
Have some overall comments. 

Also probably need someone from core to review the change to torch/_utils, @albanD who might be a reviewer here?

File: torch/nn/parallel/distributed.py (hunk 4 above)
Comment (heidongxianhua) [reply torch/nn/parallel/distributed.py:708]: yeah, there  are `XLA` test in CI flow,  I found  that the `xla` device reuses cuda-related code in distributed.py, so when the device type is xla, if we get the device module by `xla`, if will get the module named `torch.xla` which does not exist, so here I add this. @rohan-varma

**Comment by heidongxianhua** [comment]
Yeah, I have make the `device module` as a parameter for hook state. @kumpera

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py [hunk 5]
```diff
@@ -14,8 +14,9 @@ class _AllreduceUpcastHookState:
     group, and a stream to run parameter and gradient upcasts.
     """
     ddp_weakref: Any
-    upcast_stream: torch.cuda.Stream
+    upcast_stream: torch.Stream
     wait_for_stream_enqueued: bool = False
+    device_module: Any = torch.cuda
```
Comment (kumpera) [inline torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:19]: We shouldn't introduce `Any` to our typing annotations.
To address that, we should introduce a protocol that exposes the methods we need.

**Review by kumpera — CHANGES_REQUESTED** [review]

**Review by albanD — COMMENTED** [review]
The new util sounds good!

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py [hunk 6]
```diff
@@ -14,8 +14,9 @@ class _AllreduceUpcastHookState:
     group, and a stream to run parameter and gradient upcasts.
     """
     ddp_weakref: Any
-    upcast_stream: torch.cuda.Stream
+    upcast_stream: torch.Stream
```
[the remark this answers is not in the document]
Comment (heidongxianhua) [reply torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:17]: > I don't think those two types are compatible.

yeah, I have explained it at here  https://github.com/pytorch/pytorch/pull/104140#discussion_r1273097648

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py (hunk 5 above)
Comment (heidongxianhua) [reply torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py]: > We shouldn't introduce `Any` to our typing annotations. To address that, we should introduce a protocol that exposes the methods we need.

yeah, I have used protocol to do it. @kumpera

**Review by rohan-varma — COMMENTED** [review]
LGTM as long as @kumpera's concerns are addressed. Someone on distributed side should probably stamp this @kumpera @fegin @awgu

**Review by fduwjj — APPROVED** [review]
Unblock, please address the comment from @rohan-varma and @kumpera and ensure the merge conflict resolved.

**Pull request closed without merging**